.github/workflows/yocto-build-deploy.yml

```diff
@@ -0,0 +1,639 @@
+          IMAGES_PATH: ${{ github.workspace }}/build-artifacts
+        run: |
+          docker run --rm -t --user deployer \
+            -e BASE_DIR \
+            -e S3_CMD \
+            -e S3_SYNC_OPTS \
+            -e S3_BUCKET \
+            -e SLUG \
+            -e DEPLOY_ARTIFACT \
+            -e HOSTOS_VERSION \
+            -v "${IMAGES_PATH}:/host/images" \
+            "${{ steps.helper-image.outputs.imageid }}" /bin/sh -e -c '\
+              echo "${HOSTOS_VERSION}" > "/host/images/${SLUG}/latest"
+              if [ "$DEPLOY_ARTIFACT" != "docker-image" ]; then
+                /usr/src/app/node_modules/.bin/ts-node /usr/src/app/scripts/prepare.ts
+              fi
+              if [ -z "$($S3_CMD ls s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}/)" ] || [ -n "$($S3_CMD ls s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}/IGNORE)" ]; then
+                touch /host/images/${SLUG}/${HOSTOS_VERSION}/IGNORE
+                $S3_CMD del -rf s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}
+                $S3_CMD put /host/images/${SLUG}/${HOSTOS_VERSION}/IGNORE s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}/
+                $S3_CMD $S3_SYNC_OPTS dsync /host/images/${SLUG}/${HOSTOS_VERSION}/ s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}/
+                $S3_CMD put /host/images/${SLUG}/latest s3://${S3_BUCKET}/${SLUG}/ --API-ACL=public-read -f
+                $S3_CMD del s3://${S3_BUCKET}/${SLUG}/${HOSTOS_VERSION}/IGNORE
+              else
+                echo "WARNING: Deployment already done for ${SLUG} at version ${HOSTOS_VERSION}"
+              fi
+              '
+
+  ami-deploy:
+    name: AMI Deploy
+    runs-on: ${{ fromJSON(inputs.runs-on) }}
+    needs: build
+    if: inputs.deploy-ami == true
+
+    env:
+      MACHINE: "${{ inputs.machine }}"
+      BALENARC_BALENA_URL: "${{ inputs.balena-environment }}"
+      WORKSPACE: ${{ github.workspace }}
+      VERBOSE: verbose
+
+    defaults:
+      run:
+        working-directory: .
+        shell: bash --noprofile --norc -eo pipefail -x {0}
+
+    steps:
+      # TODO: clone balena-yocto-scripts @ yocto_scripts_ref instead
+      # https://github.com/actions/checkout
+      - name: Clone device repository
+        uses: actions/checkout@v3
+        with:
+          repository: ${{ inputs.device-repo || github.repository }}
+          token: ${{ secrets.GITHUB_TOKEN }}
+          ref: ${{ inputs.device-repo-ref || github.ref }}
+          submodules: true
+
+      - name: Update balena-yocto-scripts submodule to ${{ inputs.yocto-scripts-ref }}
+        if: inputs.yocto-scripts-ref != ''
+        working-directory: ./balena-yocto-scripts
+        run: |
+          git config --add remote.origin.fetch '+refs/pull/*:refs/remotes/origin/pr/*'
+          git fetch --all
+          git checkout --force "${{ inputs.yocto-scripts-ref }}"
+          git submodule update --init --recursive
+
+      # https://github.com/actions/download-artifact
+      - name: Download build artifacts
```

Comment:
Same concerns about build artifacts - they are already on the build node.

Reply:
See above answer, we are using 4 separate jobs to make use of parallelism and each job is a fresh workspace.

If this turns out to not be faster, we can revert to one job and workspace.